### PR TITLE
Add monitoring client

### PR DESCRIPTION
**Pull request by geigerj**
This PR adds generated code wrapping the gRPC monitoring library, similar to the previous PR for logging.

We have addressed some of the issues raised in the logging PR; those that have not been addressed are being tracked. We expect to have a proper github repo in the next day or so where we will be able to track those issues publicly.

### CODE REVIEW

**Comment by geigerj** [comment]
@tbetbetbe
@saicheems

**Comment by geigerj** [comment]
I realize we have a bug in our imports (lines 35 and 37 of metric_service_api.py are syntax errors). I will fix the imports and update the PR.

**Comment by geigerj** [comment]
I've updated the PR -- ready for code review now.

**Comment by dhermes** [comment]
Thanks for the heads up. Reviewing now.

File: gcloud/monitoring/agent_translation_service_api.py [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+# 1. New methods (these should be added to the end of the class)
+#
+# Happy editing!
+
+from google.gax import PageDescriptor
+from google.gax import api_callable
+from google.gax import config
+from google.gax.path_template import PathTemplate
+from google.monitoring.v3 import agent_service_pb2
+from google.api import monitored_resource_pb2
```
Comment (dhermes) [inline gcloud/monitoring/agent_translation_service_api.py]: Isn't `google.api` a core namespace? i.e. not service specific?

File: gcloud/monitoring/agent_translation_service_api.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+from google.monitoring.v3 import agent_pb2
+
+
+class AgentTranslationServiceApi(object):
+    """
+    The AgentTranslation API allows `collectd`-based agents to
```
Comment (dhermes) [inline gcloud/monitoring/agent_translation_service_api.py]: If this is intended for Sphinx, two backticks are needed:

```
``collectd``-based agents
```

File: gcloud/monitoring/agent_translation_service_api.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    write time series data to Cloud Monitoring.
+    See [google.monitoring.v3.MetricService.CreateTimeSeries] instead.
```
Comment (dhermes) [inline gcloud/monitoring/agent_translation_service_api.py]: Broken link?

File: gcloud/monitoring/agent_translation_service_api.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+    # The scopes needed to make gRPC calls to all of the methods defined in
+    # this service
+    _ALL_SCOPES = (
+        'https://www.googleapis.com/auth/monitoring',
+        'https://www.googleapis.com/auth/monitoring.write',
+        'https://www.googleapis.com/auth/monitoring.read',
+        'https://www.googleapis.com/auth/cloud-platform',
+    )
+
+
```
Comment (dhermes) [inline gcloud/monitoring/agent_translation_service_api.py]: Superfluous whitespace

File: gcloud/monitoring/agent_translation_service_api.py [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+
+    class Templates(object):
+        """PathTemplates for resources used by AgentTranslationServiceApi."""
+        PROJECT = PathTemplate.from_string(
+            'projects/{project}')
```
Comment (dhermes) [inline gcloud/monitoring/agent_translation_service_api.py]: This should be defined at module-level, though it's unclear why it is needed at all.

File: gcloud/monitoring/agent_translation_service_api.py [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.close()
+
+    def close(self):
+        del self.stub
+
+    # Service calls
+    def create_collectd_time_series(
+            self,
```
Comment (dhermes) [inline gcloud/monitoring/agent_translation_service_api.py]: Why 8-space indent here? (Sorry if this has always been that way and I just now noticed, but 4-spaces would be fine.)

File: gcloud/monitoring/agent_translation_service_api.py [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+        """
+        Creates a new time series with the given data points.  This method
+        is only for use in `collectd`-related code, including the Google
+        Monitoring Agent. See
+        [google.monitoring.v3.MetricService.CreateTimeSeries] instead.
+
+        :type name: string
+        :type resource: monitored_resource_pb2.MonitoredResource
+        :type collectd_version: string
+        :type collectd_payloads: list of agent_pb2.CollectdPayload
```
Comment (dhermes) [inline gcloud/monitoring/agent_translation_service_api.py]: These are a new feature of the codegen? Also not Google-style docstrings (though I prefer Sphinx anyhow).

File: gcloud/monitoring/agent_translation_service_api.py [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+    # Service calls
+    def create_collectd_time_series(
+            self,
+            name='',
+            resource=None,
+            collectd_version='',
+            collectd_payloads=None,
+            **kwargs):
+        """
+        Creates a new time series with the given data points.  This method
```
Comment (dhermes) [inline gcloud/monitoring/agent_translation_service_api.py]: As above in `__init__`, there is no short description here, just one long paragraph. Is this intentional? It seems these were written by hand, but maybe not?

File: gcloud/monitoring/agent_translation_service_api.py [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+            max_attempts=None):
+        return api_callable.idempotent_callable(
+            self.stub.CreateCollectdTimeSeries,
+            is_retrying=is_retrying,
+            max_attempts=max_attempts,
+            defaults=self.defaults)
+
+    # ========
+    # Manually-added methods: add custom (non-generated) methods after this point.
+    # ========
```
Comment (dhermes) [inline gcloud/monitoring/agent_translation_service_api.py]: The red line means the file doesn't end in a newline. Your files should end in a newline. (I don't know why this is the convention, but it is.)

File: gcloud/monitoring/group_service_api.py [hunk 10]
```diff
@@ -0,0 +1,342 @@
...
+from google.monitoring.v3 import group_service_pb2
+from google.api import monitored_resource_pb2
+from google.monitoring.v3 import common_pb2
+from google.monitoring.v3 import group_pb2
+
+
+class GroupServiceApi(object):
+    """
+    The Group API lets you inspect and manage your
+    [groups](google.monitoring.v3.Group).
```
Comment (dhermes) [inline gcloud/monitoring/group_service_api.py]: Broken link.

File: gcloud/monitoring/group_service_api.py [hunk 11]
```diff
@@ -0,0 +1,342 @@
...
+        'next_page_token',
+        'members',
+    )
+
+    class Templates(object):
+        """PathTemplates for resources used by GroupServiceApi."""
+        PROJECT = PathTemplate.from_string(
+            'projects/{project}')
+        GROUP = PathTemplate.from_string(
+            'projects/{project}/groups/{group}')
```
Comment (dhermes) [inline gcloud/monitoring/group_service_api.py]: Ditto here on class-in-a-class definition

File: gcloud/monitoring/group_service_api.py [hunk 12]
```diff
@@ -0,0 +1,342 @@
...
+            max_attempts=max_attempts,
+            defaults=self.defaults)
+
+    def list_groups_for_resource(
+            self,
+            name='',
+            monitored_resource=None,
+            **kwargs):
+        """
+        Lists the groups containing a specific monitored resource.
```
Comment (dhermes) [inline gcloud/monitoring/group_service_api.py]: Why not have this on the same line as the `"""`? It seems this is the pattern throughout? (I just re-read PEP257 and I think doing an entire blank line after `"""` is OK, it just looks weird to my eyes.)

File: gcloud/monitoring/group_service_api.py [hunk 13]
```diff
@@ -0,0 +1,342 @@
...
+            name=name,
+            monitored_resource=monitored_resource,
+            **kwargs)
+        return self.list_groups_for_resource_callable()(list_groups_for_resource_request)
+
+    def list_groups_for_resource_callable(
+            self,
+            is_retrying=None,
+            max_attempts=None,
+            page_streaming=_LIST_GROUPS_FOR_RESOURCE_DESCRIPTOR):
```
Comment (dhermes) [inline gcloud/monitoring/group_service_api.py]: No docstring?

Also, I get why you need an 8-space indent here, but you wouldn't (I think?) if there was a docstring.

File: gcloud/monitoring/metric_service_api.py [hunk 14]
```diff
@@ -0,0 +1,458 @@
...
+
+        :type name: string
+        :type filter: string
+        """
+
+        list_monitored_resource_descriptors_request = metric_service_pb2.ListMonitoredResourceDescriptorsRequest(
+            name=name,
+            filter=filter_,
+            **kwargs)
+        return self.list_monitored_resource_descriptors_callable()(list_monitored_resource_descriptors_request)
```
Comment (dhermes) [inline gcloud/monitoring/metric_service_api.py]: Long-line worries here and 4 lines up?

File: gcloud/monitoring/metric_service_api.py [hunk 15]
```diff
@@ -0,0 +1,458 @@
...
+        """
+        Gets a single monitored resource descriptor.
+
+        :type name: string
+        """
+
+        get_monitored_resource_descriptor_request = metric_service_pb2.GetMonitoredResourceDescriptorRequest(
+            name=name,
+            **kwargs)
+        return self.get_monitored_resource_descriptor_callable()(get_monitored_resource_descriptor_request)
```
Comment (dhermes) [inline gcloud/monitoring/metric_service_api.py]: Long line?

File: gcloud/monitoring/metric_service_api.py [hunk 16]
```diff
@@ -0,0 +1,458 @@
...
+
+    def create_metric_descriptor(
+            self,
+            name='',
+            metric_descriptor=None,
+            **kwargs):
+        """
+        Creates a new metric descriptor.
+        User-created metric descriptors define
+        [custom metrics](/monitoring/custom-metrics).
```
Comment (dhermes) [inline gcloud/monitoring/metric_service_api.py]: Link goes nowhere

**Comment by dhermes** [comment]
Finished review

**Comment by dhermes** [comment]
Did you mean to close?

**Pull request closed without merging**